Complete account upgrade and downgrade feature

praveenyr opened the pull request:
**Description :** Completed all aspects of account upgrade and downgrade feature and tested successfully.

**Question :**   I had a downgrade page link setup from the _application_html_erb.rb_ file which will invoke the _downgrade_ method in the _charges_controller.rb_.When I click the the _Downgrade_ link from the index page, the page doesnt show up but the controller method does run successfully.I suspect its got to do with the way I set up the route.

Can you please the downgrade route and let me know whats wrong with that?Also,is there is a way to call the _charges_controller/downgrade_ method directly from the _application_html_erb_ view?

**Test Coverage :** 92.57%

brittanyjspringer commented:
@praveenyr I just pulled down the code and the app is working perfectly for me. 

1. Upgrade the account
2. Click downgrade
3. Redirected to root_path, user is standard again 

I think the downgrade.html.erb and the accompanying get route are not needed. A clean POST is all the requirement needs. If you agree, please remove the extra route and view and then submit this PR into the checkpoint. Thanks!

praveenyr commented:
Brittany,
How does the application.html.erb know to invoke the downgrade method in
the charges controller? My understanding was that only view under charges/
can invoke the method.

Sorry if I'm missing something obvious.
On Sun, May 14, 2017 at 1:54 PM Brittany Martin <notifications@github.com>
wrote:

> Assigned #9 <https://github.com/praveenyr/cogniden/pull/9> to @praveenyr
> <https://github.com/praveenyr>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/praveenyr/cogniden/pull/9#event-1081403338>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ATsBogyLh6ZkLG5q_SFAwO-yO3jQfmIDks5r52ofgaJpZM4NafPp>
> .
>

brittanyjspringer commented on app/views/charges/downgrade.html.erb:
```diff
@@ -0,0 +1,4 @@
+<% if current_user.has_role? :premium %>
+    <h4>Click to downgrade to Standard user!</h4>
+    <%= link_to "Downgrade", method: :downgrade, class: 'btn btn-danger', data: { confirm: 'Are you sure you want to downgrade to standard user and relinquish power?' } %>
```
@praveenyr this line is invoking the downgrade method directly, no view needed. Since you are using current_user method in the controller, no arguments need to be passed.

praveenyr replied:
Brittany,
I meant to ask how to invoke the _downgrade_ method directly from application.html.erb. In you previous post,you did ask me to remove the get route and downgrade.html.erb, correct ?

Thanks,
Praveen.

brittanyjspringer replied:
@praveenyr I think this is where our wires are crossed -- the code you wrote is already invoking it since you are POSTing with the DELETE method. If you remove the get route and downgrade.html.erb, your code is perfect. I'm headed to roller derby practice now but will check in about 4 hours.

praveenyr replied:
Brittany,
Apologies, I'm still not getting it :(

So lets retrace my steps -

- I created this path in my layout file 
Premium User |  <%= link_to "Downgrade", downgrade_user_path %>

- I created the corresponding downgrade_user_html.erb for the same

-  I created the get route for the above 
get 'charges/downgrade' => 'charges#downgrade', :as => 'downgrade_charge'

Next,when I click on 'Downgrade' on the layout -

- It invokes a get and populates the downgrade.html.erb

- Clicking on this page will invoke a POST and call 'downgrade' method in the charges controller.

Now,when I remove the 'get' route and downgrade html,I need to invoke the 'downgrade' method in the controller directly from the layout file.And this is where I'm stuck.I cant figure out how to invoke the method without going through the route and using 'GET' to retrieve the html and then POSTing to the controller.

brittanyjspringer commented:
@praveenyr no apologies needed! I just made a code change where I removed app/views/charges/downgrade.html.erb and switched the layout view to use a POST link. 

https://github.com/praveenyr/cogniden/pull/9/commits/7fbfda3c0c9f04e71aaacfd01b42a1c5973cdfac

Can you please double check that your test suite is passing? I'm getting this result: 

Finished in 1.29 seconds (files took 3.01 seconds to load)
32 examples, 17 failures

praveenyr commented:
Brittany,
Yes,the wikis_controller_spec tests are failing.For some reason,its not able to sign in the user.Is that because the user is not set with Factorygirl here? Shouldnt matter correct? I do remember you changed this file to use FactoryGirl. Or maybe not.

praveenyr commented:
Brittany,
I fixed the downgrade route changes and tested successfully.I fixed most of the tests in the wikis controller spec but there are about 6 which are still failing.I suspect the authorization is not working on these tests and they end up getting redirected to the wrong page.Could you please help with these?

Thanks,
Praveen.

brittanyjspringer commented:
Hello, @praveenyr. Just a small change to make sure the spec user in wikis_controller_spec.rb has rights to the wiki generated. Committed. Time to merge! 

![popkey](https://m.popkey.co/8d0f9e/E8jkY.gif)
